packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -69,145 +69,241 @@ import { applyEventListeners } from './modules/events';
-// to rehydrate when dirty, because sometimes the element is not inserted just yet,
-// which breaks some invariants. For that reason, we have the following for any
-// Custom Element that is inserted via a template.
-export const ElementHook: Hooks<VElement> = {
-    create: (vnode) => {
-        const {
-            sel,
-            owner,
-            data: { svg },
-        } = vnode;
+export function patchChildren(c1: VNodes, c2: VNodes, parent: ParentNode): void {
+    if (hasDynamicChildren(c2)) {
+        updateDynamicChildren(c1, c2, parent);
+    } else {
+        updateStaticChildren(c1, c2, parent);
+    }
+}
 
-        const namespace = isTrue(svg) ? SVG_NAMESPACE : undefined;
-        const elm = createElement(sel, namespace);
-
-        linkNodeToShadow(elm, owner);
-        fallbackElmHook(elm, vnode);
-        vnode.elm = elm;
-
-        patchElementPropsAndAttrs(null, vnode);
-    },
-    update: (oldVnode, vnode) => {
-        patchElementPropsAndAttrs(oldVnode, vnode);
-        patchChildren(vnode.elm!, oldVnode.children, vnode.children);
-    },
-    insert: (vnode, parentNode, referenceNode) => {
-        insertNode(vnode, parentNode, referenceNode);
-        createChildrenHook(vnode);
-    },
-    move: insertNode,
-    remove: (vnode, parentNode) => {
-        removeNode(vnode, parentNode);
-        removeChildren(vnode);
-    },
-};
-
-export const CustomElementHook: Hooks<VCustomElement> = {
-    create: (vnode) => {
-        const { sel, owner } = vnode;
-        const UpgradableConstructor = getUpgradableConstructor(sel);
-        /**
-         * Note: if the upgradable constructor does not expect, or throw when we new it
-         * with a callback as the first argument, we could implement a more advanced
-         * mechanism that only passes that argument if the constructor is known to be
-         * an upgradable custom element.
-         */
-        let vm: VM | undefined;
-        const elm = new UpgradableConstructor((elm: HTMLElement) => {
-            // the custom element from the registry is expecting an upgrade callback
-            vm = createViewModelHook(elm, vnode);
-        });
-
-        linkNodeToShadow(elm, owner);
-        vnode.elm = elm;
-
-        if (vm) {
-            allocateChildren(vnode, vm);
-        } else if (vnode.ctor !== UpgradableConstructor) {
-            throw new TypeError(`Incorrect Component Constructor`);
-        }
-        patchElementPropsAndAttrs(null, vnode);
-    },
-    update: (oldVnode, vnode) => {
-        patchElementPropsAndAttrs(oldVnode, vnode);
-        const vm = getAssociatedVMIfPresent(vnode.elm);
-        if (vm) {
-            // in fallback mode, the allocation will always set children to
-            // empty and delegate the real allocation to the slot elements
-            allocateChildren(vnode, vm);
+function patch(n1: VNode, n2: VNode) {
+    if (n1 === n2) {
+        return;
+    }
+
```

Comment:
I would recommend adding a dev-only check here to make sure that both VNodes are the same. This will help us if we ever want to refactor the `updateDynamicChildren` method.

Reply:
Done.